skglm/solvers/gram.py

```diff
@@ -0,0 +1,272 @@
+            if ws_strategy == "subdiff":
+                opt_ws = penalty.subdiff_distance(w, grad, all_feats)
+            elif ws_strategy == "fixpoint":
+                opt_ws = dist_fix_point(w, grad, datafit, penalty, all_feats)
+
+            stop_crit = np.max(opt_ws)
+            if max(verbose - 1, 0):
+                print(f"Epoch {epoch + 1}, objective {p_obj:.10f}, "
+                      f"stopping crit {stop_crit:.2e}")
+            if stop_crit <= tol:
+                break
+            obj_out.append(p_obj)
+    return w, np.array(obj_out), stop_crit
+
+
+@njit
+def _cd_epoch_gram(G, grads, w, datafit, penalty, n_samples, n_features):
+    """Run an epoch of coordinate descent in place with gradient update using Gram.
+
+    Parameters
+    ----------
+    G : array, shape (n_features, n_features)
+        Gram matrix.
+
+    grads : array, shape (n_features,)
+        Gradient vector.
+
+    w : array, shape (n_features,)
+        Coefficient vector.
+
+    datafit : Datafit
+        Datafit.
+
+    penalty : Penalty
+        Penalty.
+
+    n_samples : int
+        Number of samples.
+
+    n_features : int
+        Number of features.
+    """
+    lc = datafit.lipschitz
+    for j in range(n_features):
+        if lc[j] == 0:
+            continue
+        old_w_j = w[j]
+        stepsize = 1 / lc[j] if lc[j] != 0 else 1000
+        w[j] = penalty.prox_1d(old_w_j + grads[j] * stepsize, stepsize, j)
+        if old_w_j != w[j]:
+            grads += G[j, :] * (old_w_j - w[j]) / n_samples
+
+
+@njit
+def _cd_epoch_gram_sparse(G_data, G_indptr, G_indices, grads, w, datafit, penalty,
+                          n_samples, n_features):
+    """Run a CD epoch with Gram update for sparse design matrices.
+
+    Parameters
+    ----------
+    G_data : array, shape (n_elements,)
+        `data` attribute of the sparse CSC matrix G.
+
+    G_indptr : array, shape (n_features + 1,)
+        `indptr` attribute of the sparse CSC matrix G.
+
+    G_indices : array, shape (n_elements,)
+        `indices` attribute of the sparse CSC matrix G.
+
+    grads : array, shape (n_features,)
+        Gradient vector.
+
+    w : array, shape (n_features,)
+        Coefficient vector.
+
+    datafit : Datafit
+        Datafit.
+
+    penalty : Penalty
+        Penalty.
+
+    n_samples : int
+        Number of samples.
+
+    n_features : int
+        Number of features.
+    """
+    lc = datafit.lipschitz
+    for j in range(n_features):
+        if lc[j] == 0:
+            continue
+        old_w_j = w[j]
+        stepsize = 1 / lc[j]
+        w[j] = penalty.prox_1d(old_w_j + grads[j] / lc[j], stepsize, j)
+        diff = old_w_j - w[j]
+        if diff != 0:
+            for i in range(G_indptr[j], G_indptr[j + 1]):
+                grads[G_indices[i]] += diff * G_data[i] / n_samples
+
+
+def fista_gram_quadratic(X, y, penalty, max_epochs=1000, tol=1e-4, w_init=None,
+                         ws_strategy="subdiff", verbose=False):
```

Comment:
there is no working set with gram so stop_crit may be better

but anyway, I propose to keep the code simple and use only subdiff_distance anyway. Would that be ok ?

Reply:
I cleaned it.